### PR TITLE
Rework load service, load path, loaded features, and autoload

### PR DESCRIPTION
**Pull request by headius**
This is a big one. This PR includes rewrites of how we handle load/require, load path lookups/caching/canonicalization, loaded feature caching, and autoload lifecycle relating to the rest.

Parts of this are loose ports from MRI, and still need some refinement and optimization. In particular, we need to make sure the load path and loaded features caches are tight, efficient, and not getting regenerated excessively. This also should be audited for concurrent state changes in autoload, since there's a few complicated interactions with threads and in-progress autoloads. We are passing most of the related specs (notably not the tags added in aea0e652866196e77b081fcfcf5f48c9c4e3752f) but the code is not proven safe in my eyes.

Hopefully others can help me audit this. I know it's a lot of code.

With this PR, here's status:

* MRI test_autoload: two excludes, for deprecation warning and a possibly bad test (threading)
* rubyspec kernel/module autoload_spec: 100%
* zeitwerk: 100%

### CODE REVIEW

**Comment by headius** [comment]
Fixes #5590.

See #5763 for mentioned perf concerns.

**Comment by headius** [comment]
Recent failure in an embedded/complete run of JRuby test suite...should be investigated for concurrency bug:

```
     [exec] ===============================================================================
     [exec] Error: test_require_fires_once_across_threads(TestRequire): RuntimeError: Concurrent requires caused double-loading
     [exec] /home/travis/build/jruby/jruby/test/jruby/test_require.rb:14:in `block in test_require_fires_once_across_threads'
     [exec]      11:       (1..(ENV['THREAD_COUNT'] || 5).to_i).map do
     [exec]      12:         Thread.new { require 'test/jruby/test_require_once_foo' }
     [exec]      13:       end.each { |t| t.join }
     [exec]   => 14:       raise "Concurrent requires caused double-loading" if $foo != 1
     [exec]      15:       $foo = 0
     [exec]      16:       $".pop
     [exec]      17:     end
     [exec] /home/travis/build/jruby/jruby/test/jruby/test_require.rb:10:in `test_require_fires_once_across_threads'
     [exec] org/jruby/RubyKernel.java:1150:in `catch'
     [exec] org/jruby/RubyKernel.java:1150:in `catch'
     [exec] ===============================================================================
```

https://travis-ci.org/jruby/jruby/jobs/544444301

**Comment by headius** [comment]
Tweaks to get `Kernel#autoload` to find the right target (6d92298be04ec237ae6615f8dc9a6d448134e0f5) appear to have regressed a few things. This is related to the "JRUBY-6570" regression spec.

**Comment by headius** [comment]
I know this is a big one...chat me up if you want to talk through some smaller pieces one at a time.

File: core/src/main/java/org/jruby/runtime/load/LibrarySearcher.java [hunk 1]
```diff
@@ -359,20 +478,58 @@ public void clear() {
...
 
-        final CharSequence key;
-        final CopyOnWriteArrayList<IRubyObject> featurePaths;
+        final StringWrapper key;
+        final List<IRubyObject> featurePaths;
     }
 
-    protected void addSingleFeatureToIndex(String key, IRubyObject featurePath) {
-        Map<String, Feature> featuresIndex = this.loadedFeaturesIndex;
+    private final ThreadLocal<StringWrapper> keyWrapper = ThreadLocal.withInitial(() -> new StringWrapper(null, 0, 0));
```
Comment (kares) [inline core/src/main/java/org/jruby/runtime/load/LibrarySearcher.java]: is this really worth the thread-local, do we allocate that many string instances on load service lookups?

**Comment by alecgorge** [comment]
@headius is there anything I can do here to help get this over the finish line? Do you need more eyes on the code or is there functionality not finished yet?

**Comment by headius** [comment]
@alecgorge The remaining work is to get autoloading with concurrent requires thread-safe. This logic was loosely ported from CRuby, which does not allow these operations to be performed in parallel, but the locking mechanisms we have available in JRuby are insufficient to make sure the case is safe. This is, I believe, the reason why the MRI suite has some sporadic failures; sometimes the parallel autoload tests work and sometimes they don't.

**Comment by rickhull** [comment]
I see the Travis CI stuff is failing -- due to Travis CI issues.  FYI, Travis CI is basically dead for open source projects.

I just did a migration to GitHub Actions for a simple repo.  I used https://bibwild.wordpress.com/2020/11/12/deep-dive-moving-ruby-projects-from-travis-to-github-actions-for-ci/ as a guide, and it was fairly straightforward.  I found my way to this issue because my tests against jruby are showing the "circular requires" warning.

**Pull request merged**